File: src/main/java/voidchess/figures/Bishop.java
package voidchess.figures;


import voidchess.board.BasicChessGameInterface;
import voidchess.board.SimpleChessBoardInterface;
import voidchess.helper.Move;
import voidchess.helper.Position;
import voidchess.image.ImageType;

import java.util.ArrayList;
import java.util.List;

import static voidchess.helper.CheckSearch.signum;

/**
 * @author stephan
 */
public class Bishop extends Figure {
    private List<Position> positions = new ArrayList<Position>(13);

    public Bishop(boolean isWhite, Position startPosition) {
        super(isWhite, startPosition, FigureType.BISHOP);
    }

    public boolean isReachable(Position to, BasicChessGameInterface game) {
        final int row_difference = to.getRow() - getPosition().getRow();
        final int column_difference = to.getColumn() - getPosition().getColumn();

        final int abs_row_difference = Math.abs(row_difference);
        final int abs_column_difference = Math.abs(column_difference);

        if (abs_row_difference != abs_column_difference || abs_row_difference == 0) {
            return false;
        }

        final int row_step = signum(row_difference);
        final int column_step = signum(column_difference);

        int row = getPosition().getRow() + row_step;
        int column = getPosition().getColumn() + column_step;

        while (row != to.getRow()) {
            final Position pos = Position.Companion.get(row, column);
            if (!game.isFreeArea(pos)) {
                return false;
            }
            row += row_step;
            column += column_step;
        }

        final Figure hitFigure = game.getFigure(to);
        return hitFigure == null || hasDifferentColor(hitFigure);
    }

    private void getNorthEastIterator(BasicChessGameInterface game, List<Position> result) {
        int row, column;

        row = getPosition().getRow();
        column = getPosition().getColumn();
        while (true) {
            row++;
            column++;
            if (row == 8 || column == 8) break;
            Position checkPosition = Position.Companion.get(row, column);
            Figure figure = game.getFigure(checkPosition);
            if (figure == null) {
                result.add(checkPosition);
            } else {
                if (hasDifferentColor(figure)) {
                    result.add(checkPosition);
                }
                break;
            }
        }
    }

    private void getSouthEastIterator(BasicChessGameInterface game, List<Position> result) {
        int row, column;

        row = getPosition().getRow();
        column = getPosition().getColumn();
        while (true) {
            row--;
            column++;
            if (row < 0 || column == 8) break;
            Position checkPosition = Position.Companion.get(row, column);
            Figure figure = game.getFigure(checkPosition);
            if (figure == null) {
                result.add(checkPosition);
            } else {
                if (hasDifferentColor(figure)) {
                    result.add(checkPosition);
                }
                break;
            }
        }
    }

    private void getNorthWestIterator(BasicChessGameInterface game, List<Position> result) {
        int row, column;

        row = getPosition().getRow();
        column = getPosition().getColumn();
        while (true) {
            row++;
            column--;
            if (row == 8 || column < 0) break;
            Position checkPosition = Position.Companion.get(row, column);
            Figure figure = game.getFigure(checkPosition);
            if (figure == null) {
                result.add(checkPosition);
            } else {
                if (hasDifferentColor(figure)) {
                    result.add(checkPosition);
                }
                break;
            }
        }
    }

    private void getSouthWestIterator(BasicChessGameInterface game, List<Position> result) {
        int row, column;

        row = getPosition().getRow();
        column = getPosition().getColumn();
        while (true) {
            row--;
            column--;
            if (row < 0 || column < 0) break;
            Position checkPosition = Position.Companion.get(row, column);
            Figure figure = game.getFigure(checkPosition);
            if (figure == null) {
                result.add(checkPosition);
            } else {
                if (hasDifferentColor(figure)) {
                    result.add(checkPosition);
                }
                break;
            }
        }
    }

    public void getReachableMoves(BasicChessGameInterface game, List<Move> result) {
        positions.clear();
        getNorthEastIterator(game, positions);
        getSouthEastIterator(game, positions);
        getNorthWestIterator(game, positions);
        getSouthWestIterator(game, positions);

        for (Position pos : positions) {
            result.add(Move.Companion.get(getPosition(), pos));
        }
    }

    public boolean isSelectable(SimpleChessBoardInterface game) {
        positions.clear();
        getNorthEastIterator(game, positions);
        for (Position pos : positions) {
            if (!isBound(pos, game)) {
                return true;
            }
        }
        positions.clear();

        getSouthEastIterator(game, positions);
        for (Position pos : positions) {
            if (!isBound(pos, game)) {
                return true;
            }
        }
        positions.clear();

        getNorthWestIterator(game, positions);
        for (Position pos : positions) {
            if (!isBound(pos, game)) {
                return true;
            }
        }
        positions.clear();

        getSouthWestIterator(game, positions);
        for (Position pos : positions) {
            if (!isBound(pos, game)) {
                return true;
            }
        }
        positions.clear();

        return false;
    }

    public int countReachableMoves(BasicChessGameInterface game) {
        positions.clear();
        getNorthEastIterator(game, positions);
        getSouthEastIterator(game, positions);
        getNorthWestIterator(game, positions);
        getSouthWestIterator(game, positions);

        return positions.size();
    }
}

File: src/main/kotlin/voidchess/figures/Bishop.kt
package voidchess.figures

import voidchess.board.BasicChessBoard
import voidchess.board.ChessBoard
import voidchess.board.check.BoundLine
import voidchess.board.check.CheckLine
import voidchess.board.move.Direction
import voidchess.board.move.Move
import voidchess.board.move.Position


class Bishop(isWhite: Boolean, startPosition: Position) : Figure(isWhite, startPosition, FigureType.BISHOP, true, false) {

    override fun isReachable(toPos: Position, game: BasicChessBoard): Boolean {
        if(!position.hasSameColor(toPos)) {
            return false
        }
        val direction = position.getDirectionTo(toPos)

        if (direction == null || direction.isStraight) {
            return false
        }

        forEachReachablePos(game, direction) {
            if (it.equalsPosition(toPos)) return true
        }

        return false
    }

    private inline fun forEachReachablePos(game: BasicChessBoard, informOf: (Position) -> Unit) {
        forEachReachablePos(game, Direction.UP_RIGHT, informOf)
        forEachReachablePos(game, Direction.UP_LEFT, informOf)
        forEachReachablePos(game, Direction.DOWN_RIGHT, informOf)
        forEachReachablePos(game, Direction.DOWN_LEFT, informOf)
    }

    override fun getReachableMoves(game: BasicChessBoard, result: MutableList<Move>) {
        forEachReachablePos(game) {
            result.add(Move[position, it])
        }
    }

    override fun getPossibleMovesWhileUnboundAndCheck(game: ChessBoard, checkLine: CheckLine, result: MutableList<Move>) {
        when {
            checkLine.posProgression.hasSinglePos -> {
                if(position.hasSameColor(checkLine.attackerPos)) {
                    addMoveIfReachable(checkLine.attackerPos, game, result)
                }
            }
            checkLine.isDiagonalCheck -> {
                if(position.hasSameColor(checkLine.attackerPos)) {
                    checkLine.posProgression.forEachReachablePos { diagonalPos->
                        if(addMoveIfReachable(diagonalPos, game, result)) {
                            // a bishop can only intersect with a diagonal attacker at a single point
                            return
                        }
                    }
                }
            }
            else -> { // isStraightCheck!
                var hasAlreadyAddedAPosition = false
                checkLine.posProgression.forEachReachablePos { straightPos->
                    if(position.hasSameColor(straightPos)) {
                        if(addMoveIfReachable(straightPos, game, result)) {
                            // a bishop can only intersect with a straight attacker at max two points
                            if(hasAlreadyAddedAPosition) return
                            else hasAlreadyAddedAPosition = true
                        }
                    }
                }
            }
        }
    }

    override fun getPossibleMovesWhileBoundAndNoCheck(game: ChessBoard, boundLine: BoundLine, result: MutableList<Move>) {
        if(boundLine.boundFigureToAttackerDirection.isDiagonal) {
            boundLine.possibleMovesToAttacker.forEachReachablePos {posBetweenThisAndAttacker->
                result.add(Move[position, posBetweenThisAndAttacker])
            }
            boundLine.possibleMovesToKing.forEachReachablePos {posBetweenThisAndKing->
                result.add(Move[position, posBetweenThisAndKing])
            }
        }
    }

    override fun isSelectable(game: ChessBoard): Boolean {
        forEachReachablePos(game) {
            if (!isBound(it, game)) return true
        }
        return false
    }

    override fun countReachableMoves(game: BasicChessBoard): Int {
        var reachableMovesCount = 0
        forEachReachablePos(game) {
            reachableMovesCount++
        }
        return reachableMovesCount
    }
}